libpromises/processes_select_proc.c

```diff
@@ -0,0 +1,809 @@
+    {
+        return false;
+    }
+
+    while (fgets(statbuf, CF_MAXVARSIZE - 1, fd))
+    {
+        sscanf(statbuf, "%63s %lu", key, &sys_boot_time);
+
+        if (strcmp(key, "btime") == 0) {
+            break;
+        }
+    }
+
+    return true;
+}
+
+/*
+ * set:
+ *   'cmd' -> ARGV[0]
+ *   'cmdline' -> space-separated ARGLIST
+ */
+static bool LoadProcCmd(pid_t pid, JsonElement *pdata)
+{
+    char statfile[CF_MAXVARSIZE];
+    FILE *fd;
+    char cmd[CF_MAXVARSIZE];
+    char *pt, *ept;
+
+    /*
+     * "/proc/.../cmdline" is "ARGV[0]\0ARGV[1]\0...ARGV[n]\0\0"
+     * For the moment we are only interested in ARGV0 (command name).
+     */
+    snprintf(statfile, CF_MAXVARSIZE, "/%s/%d/cmdline", PROCDIR, pid);
+    fd = fopen(statfile, "r");
+    if (!fd)
+    {
+        return false;
+    }
+
+    memset(cmd, '\0', CF_MAXVARSIZE);
+    pt = fgets(cmd, CF_MAXVARSIZE, fd);
+    if (pt)
+    {
+        /* key 'cmd': ARGV[0] */
+        JsonObjectAppendString(pdata, JPROC_KEY_CMD, cmd);
+
+        /*
+         * Produce full command.
+         *     Find end of 'cmdline'.
+         *     Convert intermediate '\0' to space.
+         */
+
+        /* back along buffer until ept -> last char (end of ARGV[n]) */
+        ept = &cmd[CF_MAXVARSIZE-1];
+        while (ept != cmd && *ept == '\0')
+        {
+            ept--;
+        }
+
+        /* back along buffer changing NULL to space */
+        while (ept != cmd)
+        {
+            if (*ept == '\0')
+            {
+                *ept = ' ';
+            }
+            ept--;
+        }
+
+        /* key 'cmdline': "ARGV[0] ARGV[1] ARGV[2] ... ARGV[n]" inc. spaces */
+        JsonObjectAppendString(pdata, JPROC_KEY_CMDLINE, cmd);
+    }
+    fclose(fd);
+
+    return true;
+}
+
+/*
+ * Read "/proc/pid/..." data into the "pdata" structure.
+ * There are likely to be OS dependencies.
+ *
+ * Most of the data comes from file 'stat'.
+ * But the uid is from 'status'.
+ */
+
+// Obtain and store uid/uname
+static bool LoadProcUid(JsonElement *pdata, pid_t pid)
+{
+    char statusfile[CF_MAXVARSIZE];
+
+    snprintf(statusfile, CF_MAXVARSIZE, "/%s/%d/status", PROCDIR, pid);
+
+    FILE *stream;
+    char *line = NULL;
+    size_t len = 0;
+    ssize_t nread;
+
+    stream = fopen(statusfile, "r");
+    if (stream == NULL)
+    {
+        return false;
+    }
+
+    uid_t uid;
+    bool found = false;
+    char key[CF_MAXVARSIZE];
+    while ((nread = getline(&line, &len, stream)) != -1)
+    {
+        sscanf(line, "%s %d ", key, &uid);
+        if (strcasecmp("Uid:", key) == 0)
+        {
+            found = true;
+            break;
+        }
+    }
+
+    // tidy up
+    free(line);
+    fclose(stream);
+
+    // act on file-scan result
+
+    // a "shouldn't happen" error
+    if (!found)
+    {
+        Log(LOG_LEVEL_ERR, "error searching for uid %d in %s",
+          uid, statusfile);
+
+        return false;
+    }
+
+    JsonObjectAppendInteger(pdata, JPROC_KEY_UID, uid);
```

Comment:
LGTM (code analysis) has an alert on this line, saying:

```
The variable uid may not be initialized here.
```

You say that it should be, because of the relationship with the `found` boolean variable, however there are problems. (I'll comment on relevant lines above).

Reply:
Thanks.  See next point.